llvm/lib/Transforms/Scalar/DFAJumpThreading.cpp

```diff
@@ -460,6 +461,11 @@ struct MainSwitch {
 
     BasicBlock *SIBB = SI->getParent();
 
+    // Only fold the select coming from directly where it is defined.
+    PHINode *PHIUser = dyn_cast<PHINode>(SIUse);
+    if (PHIUser && PHIUser->getIncomingBlock(*SI->use_begin()) != SIBB)
+      return false;
```

Comment:
There is a `SIBB->getSingleSuccessor() != cast<Instruction>(SIUse)->getParent()` check a few lines below -- can it be dropped now? I think it had the same purpose, but did not account for the loop entry case.